### PR TITLE
Pipelines: reproducible builds

### PR DESCRIPTION
**Pull request by scottwittenburg**
### Overview

The goal of this PR is to make gitlab pipeline builds (especially build failures) more reproducible outside of the pipeline environment.  The two key changes here which aim to improve reproducibility are: 

1. Produce a `spack.lock` during pipeline generation which is passed to child jobs via artifacts.  This concretized environment is used both by generated child jobs as well as uploaded as an artifact to be used when reproducing the build locally.
2. In the `spack ci rebuild` command, if a spec needs to be rebuilt from source, do this by generating and running an `install.sh` shell script which is then also uploaded as a job artifact to be run during local reproduction.  

To make it easier to take advantage of improved build reproducibility, this PR also adds a new subcommand, `spack ci reproduce-build`, which, given a url to job artifacts:

- fetches and unzips the job artifacts to a local directory
- looks for the generated pipeline yaml and parses it to find details about the job to reproduce
- attempts to provide a copy of the same version of spack used in the ci build
- if the ci build used a docker image, the command prints a `docker run` command you can run to get an interactive shell for reproducing the build

#### Some highlights

One consequence of this change will be much smaller pipeline yaml files.  By encoding the concrete environment in a `spack.lock` and passing to child jobs via artifacts, we will no longer need to encode the concrete root of each spec and write it into the job variables, greatly reducing the size of the generated pipeline yaml.

Additionally `spack ci rebuild` output (stdout/stderr) is no longer internally redirected to a log file, so job output will appear directly in the gitlab job trace.  With debug logging turned on, this often results in log files getting truncated because they exceed the maximum amount of log output gitlab allows.  If this is a problem, you still have the option to `tee` command output to a file in the within the artifacts directory, as now each generated job exposes a `user_data` directory as an artifact, which you can fill with whatever you want in your custom job scripts.

There are some changes to be aware of in how pipelines should be set up after this PR:

#### Pipeline generation

Because the pipeline generation job now writes a `spack.lock` artifact to be consumed by generated downstream jobs, `spack ci generate` takes a new option `--artifacts-root`, inside which it creates a `concrete_env` directory to place the lockfile.  This artifacts root directory is also where the `user_data` directory will live, in case you want to generate any custom artifacts.  If you do not provide `--artifacts-root`, the default is for it to create a `jobs_scratch_dir` within your `CI_PROJECT_DIR` (a gitlab predefined environment variable) or whatever is your current working directory if that variable isn't set. Here's the diff of the PR testing `.gitlab-ci.yml` taking advantage of the new option:

```
$ git diff develop..pipelines-reproducible-builds share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
diff --git a/share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml b/share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
index 579d7b56f3..0247803a30 100644
--- a/share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
+++ b/share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
@@ -28,10 +28,11 @@ default:
     - cd share/spack/gitlab/cloud_pipelines/stacks/${SPACK_CI_STACK_NAME}
     - spack env activate --without-view .
     - spack ci generate --check-index-only
+      --artifacts-root "${CI_PROJECT_DIR}/jobs_scratch_dir"
       --output-file "${CI_PROJECT_DIR}/jobs_scratch_dir/cloud-ci-pipeline.yml"
   artifacts:
     paths:
-      - "${CI_PROJECT_DIR}/jobs_scratch_dir/cloud-ci-pipeline.yml"
+      - "${CI_PROJECT_DIR}/jobs_scratch_dir"
   tags: ["spack", "public", "medium", "x86_64"]
   interruptible: true
```

Notice how we replaced the specific pointer to the generated pipeline file with its containing folder, the same folder we passed as `--artifacts-root`.  This way anything in that directory (the generated pipeline yaml, as well as the concrete environment directory containing the `spack.lock`) will be uploaded as an artifact and available to the downstream jobs.

#### Rebuild jobs

Rebuild jobs now must activate the concrete environment created by `spack ci generate` and provided via artifacts.  When the pipeline is generated, a directory called `concrete_environment` is created within the artifacts root directory, and this is where the `spack.lock` file is written to be passed to the generated rebuild jobs.  The artifacts root directory can be specified using the `--artifacts-root` option to `spack ci generate`, otherwise, it is assumed to be `$CI_PROJECT_DIR`.  The directory containing the concrete environment files (`spack.yaml` and `spack.lock`) is then passed to generated child jobs via the `SPACK_CONCRETE_ENV_DIR` variable in the generated pipeline yaml file.

When you don't provide custom `script` sections in your `mappings` within the `gitlab-ci` section of your `spack.yaml`, the default behavior of rebuild jobs is now to change into `SPACK_CONCRETE_ENV_DIR` and activate that environment.   If you do provide custom rebuild scripts in your `spack.yaml`, be aware those scripts should do the same thing: assume `SPACK_CONCRETE_ENV_DIR` contains the concretized environment to activate.  No other changes to existing custom rebuild scripts should be required as a result of this PR. 

As mentioned above, one key change made in this PR is the generation of the `install.sh` script by the rebuild jobs, as that same script is both run by the CI rebuild job as well as exported as an artifact to aid in subsequent attempts to reproduce the build outside of CI.  The generated `install.sh` script contains only a single `spack install` command with arguments computed by `spack ci rebuild`.  If the install fails, the job trace in gitlab will contain instructions on how to reproduce the build locally:

```
To reproduce this build locally, run:
  spack ci reproduce-build https://gitlab.next.spack.io/api/v4/projects/7/jobs/240607/artifacts [--working-dir <dir>]
If this project does not have public pipelines, you will need to first:
  export GITLAB_PRIVATE_TOKEN=<generated_token>
... then follow the printed instructions.
```

When run locally, the `spack ci reproduce-build` command shown above will download and process the job artifacts from gitlab, then print out instructions you  can copy-paste to run a local reproducer of the CI job.

This PR includes a few other changes to the way pipelines work, see the documentation on pipelines for more details.

This  PR erelies on 
~- [ ] #23194 to be able to refer to uninstalled specs by DAG hash~
EDIT: that is going to take longer to come to fruition, so for now, we will continue to install specs represented by a concrete `spec.yaml` file on disk.
- [x] #22657 to support install a single spec already present in the active, concrete environment

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@eugeneswalker @adrienbernede @shahzebsiddiqui Since I know you are all doing working with gitlab pipelines in one way or another, I wanted to draw your attention to this upcoming/proposed change to how pipelines will work.

**Comment by scottwittenburg** [comment]
@hartzell In case you're interested, this is the PR I referred to earlier on slack when I mentioned making pipeline builds more easily reproducible outside of the pipeline environment.

**Comment by scottwittenburg** [comment]
fyi @zackgalbreath

File: lib/spack/docs/pipelines.rst [hunk 1]
```diff
@@ -490,13 +629,15 @@ Using a custom spack in your pipeline
...
+user-provided pipeline scripts to accomplish this fairly simply.  First, consider
+specifying the source and version of spack you want to use with variables, either
+written directly into your ``.gitlab-ci.yml``, or provided by CI variables defined
+in the gitlab UI or from some upstream pipeline.  Let's say you choose the variable
+names ``SPACK_REPO`` and ``SPACK_REF`` to refer to the particular fork of spack
+and branch you want for running your pipeline.  You can then refer to those in a
+custom shell script invoked both from your pipeline generation job and your rebuild
 jobs.  Here's the ``generate-pipeline`` job from the top of this document,
-updated to invoke a custom shell script that will clone and source a custom
+updated to clone and source a custom
```
Comment (becker33) [inline lib/spack/docs/pipelines.rst]: Source a custom what?

File: lib/spack/docs/pipelines.rst [hunk 2]
```diff
@@ -505,34 +646,24 @@ spack:
      tags:
        - <some-other-tag>
    before_script:
-     - ./cloneSpack.sh
+     - git clone ${SPACK_REPO}
+     - pushd spack && git checkout ${SPACK_REF} && popd
```
Comment (becker33) [inline lib/spack/docs/pipelines.rst:649]: Shouldn't this technically be be `pushd spack && git checkout ${SPACK_REF}; popd` so that we properly fail if the `git checkout` command fails?

File: lib/spack/docs/pipelines.rst [hunk 3]
```diff
@@ -547,17 +678,17 @@ of spack, so you would update your ``spack.yaml`` from above as follows:
                - spack-kube
              image: spack/ubuntu-bionic
              before_script:
-               - ./cloneSpack.sh
+               - git clone ${SPACK_REPO}
+               - pushd spack && git checkout ${SPACK_REF} && popd
```
Comment (becker33) [inline lib/spack/docs/pipelines.rst:681]: Same as above

File: lib/spack/docs/pipelines.rst [hunk 4]
```diff
@@ -594,28 +744,33 @@ environment variables used by the pipeline infrastructure are described here.
 AWS_ACCESS_KEY_ID
 ^^^^^^^^^^^^^^^^^
 
-Needed when binary mirror is an S3 bucket.
+Optional.  Only needed when binary mirror is an S3 bucket.
```
Comment (becker33) [inline lib/spack/docs/pipelines.rst:746]: Are the triple-spaces intentional?

File: lib/spack/spack/ci.py [hunk 5]
```diff
@@ -513,28 +513,31 @@ def format_job_needs(phase_name, strip_compilers, dep_jobs,
...
-                            check_index_only=False, run_optimizer=False,
-                            use_dependencies=False):
+def generate_gitlab_ci_yaml(env, print_summary, output_file,
+                            prune_dag=False, check_index_only=False,
+                            run_optimizer=False, use_dependencies=False,
+                            artifacts_root=None):
     # FIXME: What's the difference between one that opens with 'spack'
     # and one that opens with 'env'?  This will only handle the former.
     with spack.concretize.disable_compiler_existence_check():
-        env.concretize()
```
Comment (becker33) [inline lib/spack/spack/ci.py:537]: I think the comment above this code block is not true anymore (since we use `ev.config_dict` below)

File: lib/spack/spack/ci.py [hunk 6]
```diff
@@ -513,28 +513,31 @@ def format_job_needs(phase_name, strip_compilers, dep_jobs,
...
+    items.insert(0, (name, url))
     cfg.set('mirrors', syaml.syaml_dict(items), scope=cfg_scope)
 
 
-def remove_pr_mirror():
+def remove_mirror(name):
     cfg_scope = cfg.default_modify_scope()
     mirrors = cfg.get('mirrors', scope=cfg_scope)
-    mirrors.pop('ci_pr_mirror')
+    mirrors.pop(name)
```
Comment (becker33) [inline lib/spack/spack/ci.py]: These methods should move to `spack.mirror` and should be called by the `add_mirror`  and `remove_mirror` methods in `spack.cmd.mirror` to reduce code duplication.

File: lib/spack/spack/ci.py [hunk 7]
```diff
@@ -1335,3 +1404,278 @@ def copy_stage_logs_to_artifacts(job_spec, job_log_dir):
...
+    st = os.stat(install_script)
+    os.chmod(install_script, st.st_mode | stat.S_IEXEC)
+
+    # Find the repro details file
+    repro_file = fs.find(work_dir, 'repro.json')[0]
+    repro_details = None
+    with open(repro_file) as fd:
+        repro_details = json.load(fd)
+
+    mount_as_dir = '/work'
```
Comment (becker33) [inline lib/spack/spack/ci.py]: This is really hard  to parse  mentally before reading the entire method. I think it's fine as-is, we just need comments so other developers can  figure out what's happening.

```
# construct directory for reproducer on container image
```
would go a long way towards making this easier to follow.

That leads to a second point -- I think this would be easier to follow if we got the `job_image` variable before this, and only compute the `mount_as_dir` if we actually need it.

In general, the rest of this method is hard to parse, and per-stanza comments can probably resolve the whole issue.

File: lib/spack/spack/cmd/ci.py [hunk 8]
```diff
@@ -245,208 +263,314 @@ def ci_rebuild(args):
...
 
-    with open(job_log_file, 'w') as log_fd:
-        os.dup2(log_fd.fileno(), sys.stdout.fileno())
-        os.dup2(log_fd.fileno(), sys.stderr.fileno())
+    # Check configured mirrors for a built spec with a matching full hash
+    matches = bindist.get_mirrors_for_spec(
+        job_spec, full_hash_match=True, index_only=False)
+
+    if matches:
+        # Got at full hash match on at least one configured mirror.  All
```
Comment (becker33) [inline lib/spack/spack/cmd/ci.py]:
```suggestion
        # Got a full hash match on at least one configured mirror.  All
```

File: lib/spack/spack/cmd/ci.py [hunk 9]
```diff
@@ -16,14 +20,16 @@
...
-import spack.util.executable as exe
 import spack.util.url as url_util
+import spack.util.web as web_util
 
 
 description = "manage continuous integration pipelines"
 section = "build"
 level = "long"
 
+CI_REBUILD_INSTALL_BASE_ARGS = ['-d', '-k', '-v']
```
Comment (becker33) [inline lib/spack/spack/cmd/ci.py]: Why are we turning off SSL checking  for all CI rebuilds? Shouldn't we pick that up from the Spack configuration and pass it along to the subprocess?

File: lib/spack/spack/cmd/ci.py [hunk 10]
```diff
@@ -245,208 +263,314 @@ def ci_rebuild(args):
...
+
+        # store the cdash build id on disk for later
+        with open(cdash_id_path, 'w') as fd:
+            fd.write(cdash_build_id)
+
+    # A compiler action of 'FIND_ANY' means we are building a bootstrap
+    # compiler or one of its deps.
+    # TODO: when compilers are dependencies, we should include --no-add
+    if compiler_action != 'FIND_ANY':
+        install_args.append('--no-add')
```
Comment (becker33) [inline lib/spack/spack/cmd/ci.py:447]: to confirm: The compiler bootstrapping specs that get added to the yaml are only added to the local yaml for this job, right?

File: lib/spack/spack/cmd/ci.py [hunk 11]
```diff
@@ -245,208 +263,314 @@ def ci_rebuild(args):
...
+    export GITLAB_PRIVATE_TOKEN=<generated_token>
+
+... then follow the printed instructions.
+
+""".format(repro_job_url)
+
+        print(reproduce_msg)
+
+    # Tie job success/failure to the success/failure of building the spec
+    sys.exit(install_exit_code)
```
Comment (becker33) [inline lib/spack/spack/cmd/ci.py:573]: +1

File: lib/spack/spack/test/ci.py [hunk 12]
```diff
@@ -88,30 +91,42 @@ def assert_present(config):
...
-        nonconc_root_spec, 'archive-files', dep_builds, 'FIND_ANY')
+        concrete_deps = spec_map['deps']
+        for key, obj in concrete_deps.items():
+            assert(obj.build_hash() == hash_dict[key])
 
-    assert('root' in spec_map and 'deps' in spec_map)
-    assert('archive-files' in spec_map)
+        s = spec.Spec('dyninst')
+        nonconc_root_spec = 'dyninst'
+        print('nonconc spec name: {0}'.format(nonconc_root_spec))
```
Comment (becker33) [inline lib/spack/spack/test/ci.py]: Couldn't you just use `s.name` instead of a separate variable here?

File: lib/spack/spack/test/ci.py [hunk 13]
```diff
@@ -126,32 +141,39 @@ def test_register_cdash_build():
...
+        cdash_project = 'spack'
+        cdashids_mirror_url = 'https://my.fake.mirror'
+
+        with pytest.raises(spack.error.SpackError):
+            ci.relate_cdash_builds(spec_map, cdash_api_url, job_build_id,
+                                   cdash_project, [cdashids_mirror_url])
+
+        # Just make sure passing None for build id doesn't throw exceptions
+        ci.relate_cdash_builds(spec_map, cdash_api_url, None, cdash_project,
+                               [cdashids_mirror_url])
```
Comment (becker33) [inline lib/spack/spack/test/ci.py]: Is all this test does confirming that it fails for a missing url and succeeds for `None`?

File: lib/spack/spack/test/cmd/ci.py [hunk 14]
```diff
@@ -704,27 +714,111 @@ def test_ci_rebuild_basic(tmpdir, mutable_mock_env_path, env_deactivate,
...
+        print('  {0}'.format(f))
+
+    site_config_yaml_path = os.path.join(site_config_path, 'config.yaml')
+    with open(site_config_yaml_path) as fd:
+        print('config.yaml contents:')
+        print(fd.read())
+
+    # test_configs = os.path.join(spack_paths.test_path, 'data', 'config')
+    monkeypatch.setattr(spack.cmd.ci, 'CI_REBUILD_INSTALL_BASE_ARGS', [
+        '-d', '-k', '-v', '-m', '-C', site_config_path
```
Comment (becker33) [inline lib/spack/spack/test/cmd/ci.py]: Have you confirmed that this appropriately mocks the install_tree? That's  the only thing  I can think  of that I don't see how it's being mocked at this point.

**Review by becker33 — DISMISSED** [review]
Some minor requests, mostly documentation,  and a few questions  to confirm other sections are doing what I think they are.

File: lib/spack/docs/pipelines.rst (hunk 2 above)
Comment (scottwittenburg) [reply lib/spack/docs/pipelines.rst:649]: The double-amp is a short-circuit logical operator, isn't it?   So if the checkout fails, the popd isn't run and the whole line fails?  Maybe I have that wrong.

File: lib/spack/docs/pipelines.rst (hunk 4 above)
Comment (scottwittenburg) [reply lib/spack/docs/pipelines.rst:746]: Those are double spaces, is there a convention about that in `rst`?  Should we just insert a single space after a period?

File: lib/spack/spack/cmd/ci.py (hunk 10 above)
Comment (scottwittenburg) [reply lib/spack/spack/cmd/ci.py:447]: Yes, that is the way it is implemented currently.

File: lib/spack/docs/pipelines.rst (hunk 2 above)
Comment (becker33) [reply lib/spack/docs/pipelines.rst:649]: You are correct, I was thinking of `||` when I thought through the return codes.

File: lib/spack/docs/pipelines.rst (hunk 4 above)
Comment (becker33) [reply lib/spack/docs/pipelines.rst:746]: I could've sworn I counted 3, but apparently I've learned how to count since 10am this morning. That looks fine (I don't like the double-space after a period, but I know I'm on the losing side of that).

File: lib/spack/docs/pipelines.rst (hunk 1 above)
Comment (scottwittenburg) [reply lib/spack/docs/pipelines.rst]: Whoops, there was a newline there before the word "spack".  Fixed.

File: lib/spack/spack/test/ci.py (hunk 12 above)
Comment (scottwittenburg) [reply lib/spack/spack/test/ci.py]: Fixed

File: lib/spack/spack/test/ci.py (hunk 13 above)
Comment (scottwittenburg) [reply lib/spack/spack/test/ci.py]: Yeah, that was kind of useless.  I fixed this and a few other tests to mock `build_opener` from `six.moves.urllib.request` to return something I can use to actually check things in methods that make http requests.

File: lib/spack/spack/test/cmd/ci.py (hunk 14 above)
Comment (scottwittenburg) [reply lib/spack/spack/test/cmd/ci.py]: This is the last thing I need to look into, which I'm doing now.  But if you don't have any other requested changes, maybe this doesn't have to hold up the PR.

File: lib/spack/spack/test/cmd/ci.py [hunk 15]
```diff
@@ -1456,3 +1550,122 @@ def test_ci_generate_read_broken_specs_url(tmpdir, mutable_mock_env_path,
...
+    rep_out = ci_cmd('reproduce-build',
+                     'https://some.domain/api/v1/projects/1/jobs/2/artifacts',
+                     '--working-dir',
+                     working_dir.strpath,
+                     output=str)
+
+    expect_out = 'docker run --rm -v {0}:{0} -ti {1}'.format(
+        working_dir.strpath, image_name)
+
+    print('spack ci reproduce-build ...')
```
Comment (becker33) [inline lib/spack/spack/test/cmd/ci.py]: It looks like there are still some debug print statements in this test

File: lib/spack/spack/test/ci.py (hunk 13 above)
Comment (scottwittenburg) [reply lib/spack/spack/test/ci.py]: My comment was intended to indicate I think I've resolved this, as this test (and a few others).  I did leave in this one call that passes a build_id of `None` because there's a check for that at the top of the method, and I wanted that line covered.  But notice that test now makes two calls to the same method above that actually make useful assertions.

File: lib/spack/spack/test/cmd/ci.py (hunk 15 above)
Comment (scottwittenburg) [reply lib/spack/spack/test/cmd/ci.py]: In general I don't try to remove those since they don't get printed unless there's a failure.  If it's important to remove this, then should I also go through and remove all the other print statements from this module?

Comment (becker33) [reply lib/spack/spack/test/cmd/ci.py]: I guess it doesn't matter -- I've always made sure to remove them since it's not part  of the test proper.

File: lib/spack/spack/test/cmd/ci.py (hunk 14 above)
Comment (becker33) [reply lib/spack/spack/test/cmd/ci.py]: This looks good now.

File: lib/spack/spack/cmd/ci.py (hunk 9 above)
Comment (scottwittenburg) [reply lib/spack/spack/cmd/ci.py]: That's a holdover from early days when it was required to get things to build.  I agree it should come from configs.

**Review by tgamblin — DISMISSED** [review]

**Review by tgamblin — APPROVED** [review]
